ledger/acctupdates.go

```diff
@@ -288,12 +325,253 @@ func (au *accountUpdates) close() {
+		}
+
+	}
+}
+
+func (au *accountUpdates) LookupKeysByPrefix(round basics.Round, keyPrefix string, maxKeyNum uint64) ([]string, error) {
+	return au.lookupKeysByPrefix(round, keyPrefix, maxKeyNum, true /* take lock */)
+}
+
+func (au *accountUpdates) lookupKeysByPrefix(round basics.Round, keyPrefix string, maxKeyNum uint64, synchronized bool) (resultKeys []string, err error) {
+	var results map[string]bool
+	// keep track of the number of result key with value
+	var resultCount uint64
+
+	needUnlock := false
+	if synchronized {
+		au.accountsMu.RLock()
+		needUnlock = true
+	}
+	defer func() {
+		if needUnlock {
+			au.accountsMu.RUnlock()
+		}
+		// preparation of result happens in deferring function
+		// prepare result only when err != nil
+		if err == nil {
+			resultKeys = make([]string, 0, resultCount)
+			for resKey, present := range results {
+				if present {
+					resultKeys = append(resultKeys, resKey)
+				}
+			}
+		}
+	}()
+
+	// TODO: This loop and round handling is copied from other routines like
+	// lookupResource. I believe that it is overly cautious, as it always reruns
+	// the lookup if the DB round does not match the expected round. However, as
+	// long as the db round has not advanced too far (greater than `rnd`), I
+	// believe it would be valid to use. In the interest of minimizing changes,
+	// I'm not doing that now.
+
+	for {
+		currentDBRound := au.cachedDBRound
+		currentDeltaLen := len(au.deltas)
+		offset, _err := au.roundOffset(round)
+		if _err != nil {
+			err = _err
+			return
+		}
+
+		// reset `results` to be empty each iteration
+		// if db round does not match the round number returned from DB query, start over again
+		// NOTE: `results` is maintained as we walk backwards from the latest round, to DB
+		// IT IS NOT SIMPLY A SET STORING KEY NAMES!
+		// - if the boolean for the key is true: we consider the key is still valid in later round
+		// - otherwise, we consider that the key is deleted in later round, and we will not return it as part of result
+		// Thus: `resultCount` keeps track of how many VALID keys in the `results`
+		// DO NOT TRY `len(results)` TO SEE NUMBER OF VALID KEYS!
+		results = map[string]bool{}
+		resultCount = 0
+
+		for i := int(offset - 1); i >= 0; i-- {
+			for keyInRound, mv := range au.kvDeltas[i] {
+				if !strings.HasPrefix(keyInRound, keyPrefix) {
+					continue
+				}
+				// whether it is set or deleted in later round, if such modification exists in later round
+				// we just ignore the earlier insert
+				if _, ok := results[keyInRound]; ok {
+					continue
+				}
+				if mv.Data == nil {
+					results[keyInRound] = false
+				} else {
+					// set such key to be valid with value
+					results[keyInRound] = true
+					resultCount++
+					// check if the size of `results` reaches `maxKeyNum`
+					// if so just return the list of keys
+					if maxKeyNum > 0 && resultCount == maxKeyNum {
```

Comment:
did you consider `maxKeyNum = MaxUint64` to indicate "no limit"? In this case we can remove all the `maxKeyNum > 0` checks

Reply:
I think that's very reasonable.  That's how listCreatables works. @ahangsu would it be a big change? Basically coerce 0 to MaxUint64 higher up (handlers.go?) so this interface is cleaner - max really means max.

Reply:
Lifting the condition `maxKeyNum > 0` is not too much a big deal to me, for `maxKeyNum` means returning all keys, all of them. This is really meaning 0 is $\infty$ num of keys.

The `resultCount` on line 521 can only be greater than 0, and lifting such condition is fine, for it never equal 0, but the interface is just straightforward (for me I guess?).

Reply:
Resolving - Addressed in https://github.com/algorand/go-algorand/pull/4681 by @jannotti.